gps/deduce.go

```diff
@@ -167,36 +169,74 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
 	return "bitbucket.org" + v[2], nil
 }
 
+// getBitbucketRepo expects the input to be the path after the bitbucket.org
+// hostname in an import URL. The following inputs should be valid:
+//
+// 		/user/repo
+// 		/user/repo.git
+// 		/user/repo/some/package
+// 		/user/repo.git/some/package
+func getBitbucketRepo(s string) (string, error) {
+	ss := strings.Split(s, "/")
+	if len(ss) < 3 {
+		return "", errors.Errorf("bitbucket repo path %q appears malformed, want /USER/REPO", s)
+	}
+
+	repoStr := ss[1] + "/" + ss[2]
+
+	if strings.HasSuffix(repoStr, ".git") {
+		repoStr = repoStr[:len(repoStr)-4]
+	}
+
+	if strings.HasSuffix(repoStr, ".hg") {
+		repoStr = repoStr[:len(repoStr)-3]
+	}
+
+	return repoStr, nil
+}
+
 func (m bitbucketDeducer) deduceSource(path string, u *url.URL) (maybeSources, error) {
 	v := m.regexp.FindStringSubmatch(path)
 	if v == nil {
 		return nil, fmt.Errorf("%s is not a valid path for a source on bitbucket.org", path)
 	}
 
+	bbRepo, err := getBitbucketRepo(v[2])
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to extract repo from bitbucket URL")
+	}
+
+	// the calls we are making are for repository metadata, which should be
+	// pretty quick.
+	//
+	// this timeout value is totally arbitrary, but should be within normal
+	// bounds
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+
+	scm, err := bitbucketapi.GetSCM(ctx, bbRepo)
+	cancel()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get SCM details from bitbucket repo")
+	}
+
+	isgit := scm == "git"
+
 	u.Host = "bitbucket.org"
 	u.Path = v[2]
 
-	// This isn't definitive, but it'll probably catch most
-	isgit := strings.HasSuffix(u.Path, ".git") || (u.User != nil && u.User.Username() == "git")
-	ishg := strings.HasSuffix(u.Path, ".hg") || (u.User != nil && u.User.Username() == "hg")
-
-	// TODO(sdboyer) resolve scm ambiguity if needed by querying bitbucket's REST API
 	if u.Scheme != "" {
 		validgit, validhg := validateVCSScheme(u.Scheme, "git"), validateVCSScheme(u.Scheme, "hg")
 		if isgit {
 			if !validgit {
-				// This is unreachable for now, as the git schemes are a
-				// superset of the hg schemes
 				return nil, fmt.Errorf("%s is not a valid scheme for accessing a git repository", u.Scheme)
 			}
 			return maybeSources{maybeGitSource{url: u}}, nil
-		} else if ishg {
+		} else if !isgit {
```

Comment:
We can just say `else` here since this covers the exact opposite case as the first block.